Upgrade to xenial

psivesely opened the pull request:
Upgrading our Vagrant image to Xenial would mean that (i) we wouldn't have to rely on 3rd party PPAs for Python 3.5 and (ii) that we'd have access to pip3.5, which would resolve the need to use the inflexible `-t` method with pip (and also would allow us to use the Ansible `pip` module instead of shelling out wityh `command`).

conorsch committed:
Updates base box trusty64 -> xenial64

Since we're using recent versions of Python (3.5+), let's switch to
Ubuntu 16.04 for the base box, rather than shoehorning custom PPAs into
the older 14.04 system.

There are a few gotchas with shared folder support for Vagrant under
Xenial, thus the hacks.

conorsch committed:
Updates vagrant check to catch Ubuntu 16.04

The username "vagrant" is standard for VirtualBox vagrant images, but
Ubuntu 16.04 sets the default username to "ubuntu", so let's check both
and skip cloning if either is found.

conorsch committed:
Removes PPA for installing Python 3.5

Xenial image has Python 3.5.1 installed by default, so no need to add a
custom PPA strictly for more recent versions of Python.

conorsch committed:
Consolidates apt tasks in common role

Using Xenial allows us to simplify the apt packages, since Python3.5 is
available by default.

conorsch committed:
Uses system-wide config file for tmux

No need to write to each user's home dir to set a default. Writing to
the system directory allows per-user overrides, as well.

conorsch committed:
Uses pip module for installing Python dependencies

That's what the pip module is for! The role used 'command' previously as
a hack to get around multiple Python versions, but since the system
default in Xenial is Python3.5, we simply need to specify 'pip3' for the
'executable' parameter, and Ansible will handle the rest.

conorsch committed:
Adds helpful error message to first-run provision

Since the shared folder won't be configured on the first 'up', due to
regressions in the Xenial Vagrant box [0], let's detect its absence
and print out instructions for the developer. #docops \m/

[0] https://bugs.launchpad.net/cloud-images/+bug/1565985

psivesely commented on roles/crawler/tasks/install-python-deps.yml:
```diff
@@ -1,6 +1,6 @@
-  args:
-    chdir: "~{{ ssh_username }}/FingerprintSecureDrop/fpsd/requirements"
+  pip:
+    executable: pip3
```
This should be pip3.5

conorsch replied:
No, pip3 is correct:

```
ubuntu@ubuntu-xenial:~$ which pip3.5
ubuntu@ubuntu-xenial:~$ pip3 --version
pip 8.1.1 from /usr/lib/python3/dist-packages (python 3.5)
```

conorsch commented:
Due to a regression in the Ubuntu Vagrant boxes, the shared folder functionality is rather broken: https://bugs.launchpad.net/cloud-images/+bug/1565985 Therefore we'll resort to a two-step provisioning run on box creation. Added an informative message so folks know what to do:

```
==> default: You must manually enable the shared folder in the Vagrantfile.
==> default: Edit the 'config.vm.synced_folder' line and set 'disabled' to 'false',
==> default: then run 'vagrant reload --provision' to reboot and mount the shared folder.
==> default: We do apologize for the inconvenience. Upstream changes are pending
==> default: to avoid this manual bootstrapping process in the future.
```

Once the upstream box is patched, we can remove the custom scripting logic. From the discussion in that thread, looks like a fix might be released in a few weeks.

garrettr commented on roles/common/tasks/main.yml:
```diff
@@ -1,4 +1,27 @@
+    state: present
+    update_cache: yes
+    # Only update if it's been 24 hours since last time
+    cache_valid_time: 86400
```
Cool, I didn't know about this before.

conorsch commented:
Not quite there yet—we're still missing some pip dependencies:

```
ubuntu@ubuntu-xenial:~$ cd FingerprintSecureDrop/fpsd
ubuntu@ubuntu-xenial:~/FingerprintSecureDrop/fpsd$ ./sorter.py 
ubuntu@ubuntu-xenial:~/FingerprintSecureDrop/fpsd$ ./crawler.py 
Traceback (most recent call last):
  File "./crawler.py", line 28, in <module>
    from tbselenium.tbdriver import TorBrowserDriver
ImportError: No module named 'tbselenium'
```

conorsch committed:
Uses pip module for python3 dependencies in sorter role

Same as we're now doing in the crawler role, we can use the Ansible
'pip' module with the 'pip3' executable parameter to install Python
packages system-wide. Xenial defaults to Python 3.5.1, so we don't need
to muck around with the -t flag to pip.
